File: web/src/api/prometheus.ts
import { api } from '@/shared/api/client'

export type PrometheusAuthType = 'none' | 'basic' | 'bearer'
export type PrometheusScopeType = 'topology_layer' | 'host'

export interface PrometheusSource {
  id: string
  name: string
  base_url: string
  timeout_seconds: number
  auth_type: PrometheusAuthType
  username?: string
  skip_tls_verify: boolean
  created_at: string
  updated_at: string
}

export interface PrometheusBinding {
  id: string
  source_id: string
  scope_type: PrometheusScopeType
  topology_id?: string
  layer?: string
  host_id?: string
  created_at: string
}

export interface AddPrometheusSourceRequest {
  name: string
  base_url: string
  timeout_seconds?: number
  auth_type: PrometheusAuthType
  username?: string
  password?: string
  token?: string
  skip_tls_verify?: boolean
}

export interface UpdatePrometheusSourceRequest {
  name?: string
  base_url?: string
  timeout_seconds?: number
  auth_type?: PrometheusAuthType
  username?: string
  password?: string
  token?: string
  skip_tls_verify?: boolean
}

export async function listPrometheusSources(): Promise<PrometheusSource[]> {
  return api.get<PrometheusSource[]>('/prometheus/sources')
}

export async function addPrometheusSource(req: AddPrometheusSourceRequest): Promise<PrometheusSource> {
  return api.post<PrometheusSource>('/prometheus/sources', req)
}

export async function updatePrometheusSource(id: string, req: UpdatePrometheusSourceRequest): Promise<PrometheusSource> {
  return api.put<PrometheusSource>(`/prometheus/sources/${id}`, req)
}

export async function deletePrometheusSource(id: string): Promise<void> {
  return api.delete<void>(`/prometheus/sources/${id}`, { responseType: 'void' })
}

export async function testPrometheusConnection(id: string): Promise<{ ok: boolean; latency_ms?: number; error?: string }> {
  return api.get<{ ok: boolean; latency_ms?: number; error?: string }>(`/prometheus/sources/${id}/test`)
}

export async function addPrometheusBinding(req: {
  source_id: string
  scope_type: PrometheusScopeType
  topology_id?: string
  layer?: string
  host_id?: string
}): Promise<PrometheusBinding> {
  return api.post<PrometheusBinding>('/prometheus/bindings', req)
}

export async function deletePrometheusBinding(id: string): Promise<void> {
  return api.delete<void>(`/prometheus/bindings/${id}`, { responseType: 'void' })
}

File: internal/models/prometheus.go
package models

import "time"

type PrometheusAuthType string

const (
	PrometheusAuthNone   PrometheusAuthType = "none"
	PrometheusAuthBasic  PrometheusAuthType = "basic"
	PrometheusAuthBearer PrometheusAuthType = "bearer"
)

type PrometheusSource struct {
	ID                string             `json:"id"`
	Name              string             `json:"name"`
	BaseURL           string             `json:"base_url"`
	TimeoutSeconds    int                `json:"timeout_seconds"`
	AuthType          PrometheusAuthType `json:"auth_type"`
	Username          string             `json:"username,omitempty"`
	EncryptedPassword string             `json:"-"`
	EncryptedToken    string             `json:"-"`
	SkipTLSVerify     bool               `json:"skip_tls_verify"`
	CreatedAt         time.Time          `json:"created_at"`
	UpdatedAt         time.Time          `json:"updated_at"`
}

type PrometheusScopeType string

const (
	ScopeTopologyLayer PrometheusScopeType = "topology_layer"
	ScopeHost          PrometheusScopeType = "host"
)

type PrometheusBinding struct {
	ID         string              `json:"id"`
	SourceID   string              `json:"source_id"`
	ScopeType  PrometheusScopeType `json:"scope_type"`
	TopologyID string              `json:"topology_id,omitempty"`
	Layer      string              `json:"layer,omitempty"`
	HostID     string              `json:"host_id,omitempty"`
	CreatedAt  time.Time           `json:"created_at"`
}

type AddPrometheusSourceRequest struct {
	Name           string             `json:"name"`
	BaseURL        string             `json:"base_url"`
	TimeoutSeconds int                `json:"timeout_seconds"`
	AuthType       PrometheusAuthType `json:"auth_type"`
	Username       string             `json:"username"`
	Password       string             `json:"password"`
	Token          string             `json:"token"`
	SkipTLSVerify  bool               `json:"skip_tls_verify"`
}

type UpdatePrometheusSourceRequest struct {
	Name           *string             `json:"name"`
	BaseURL        *string             `json:"base_url"`
	TimeoutSeconds *int                `json:"timeout_seconds"`
	AuthType       *PrometheusAuthType `json:"auth_type"`
	Username       *string             `json:"username"`
	Password       *string             `json:"password"`
	Token          *string             `json:"token"`
	SkipTLSVerify  *bool               `json:"skip_tls_verify"`
}

type AddPrometheusBindingRequest struct {
	SourceID   string              `json:"source_id"`
	ScopeType  PrometheusScopeType `json:"scope_type"`
	TopologyID string              `json:"topology_id"`
	Layer      string              `json:"layer"`
	HostID     string              `json:"host_id"`
}
